src/main/host/memory_manager.rs

```diff
@@ -344,8 +362,24 @@ where
     }
 
     fn read_some(&self, offset: usize, buf: &mut [T]) -> Result<usize, Errno> {
-        self.memory_manager
-            .readv_ptrs(&mut [buf], &[self.ptr.slice(offset..)])
+        // Split at page boundaries to allow partial reads.
+        let mut slices = Vec::with_capacity((buf.len() + page_size() - 1) / page_size());
+        let mut toread = buf.len();
+        let mut buf = buf;
+        while toread > 0 {
+            let mut this_toread = toread % page_size();
+            if this_toread == 0 {
+                this_toread = page_size();
+            }
+            let (prefix, suffix) = buf.split_at_mut(this_toread);
+            buf = suffix;
+            slices.push(prefix);
+            toread -= this_toread;
+        }
+        let rv = self
+            .memory_manager
+            .readv_ptrs(&mut slices, &[self.ptr.slice(offset..)]);
+        rv
```

Comment:
1. This seems like it's mixing counts (`buf.len()`) with bytes (`page_size()`). Should this be using `buf.len() * sizeof(T)`?
2. This seems to be reading in chunks of `page_size()`, but the chunks don't seem to be aligned at page boundaries if `self.ptr` itself isn't aligned to a page boundary.